### PR TITLE
Fix login rate limiting

### PR DESCRIPTION
**Pull request by taquitos**
implemented new endpoints for login
Fix for #10793, also #10812

### CODE REVIEW

**Review by KrauseFx — APPROVED** [review]

File: spaceship/lib/spaceship/client.rb
Comment (casz) [inline spaceship/lib/spaceship/client.rb]: How did we arrive on this magic appIdKey ❓

Comment (taquitos) [reply spaceship/lib/spaceship/client.rb]: Great question, I cleared my history and navigated to developer.apple.com
then I clicked on `account` which took me to the login page `https://idmsa.apple.com/IDMSWebAuth/login?appIdKey=891bd3417a7776362562d2197f89480a8547b108fd934911bcbea0110d07f757&path=%2Faccount%2F&rv=1`
I then tracked the `POST` request and that `appIdKey` was included.
I resent the request without the appIdKey and it failed, so I added it back in and the request succeeded.
My guess is that Apple is using SSO now, and each service they operate will have an "appIdKey" associated with it.

Comment (taquitos) [reply spaceship/lib/spaceship/client.rb]: We might be able to derive this `appIdKey` from somewhere, but I didn't find it yet.

**Review by joshdholtz — APPROVED** [review]
Might be good to test with 2FA before this 🚢 s but code looks 👌

**Review by jetersen — APPROVED** [review]
🚀

**Comment by taquitos** [comment]
I'm going to do some 2FA testing real quick

File: spaceship/lib/spaceship/client.rb
Comment (lacostej) [reply spaceship/lib/spaceship/client.rb]: If you google for it, you will find plenty of references on other's people scripts :)

**Comment by taquitos** [comment]
2FA doesn't work, looking into that

**Review by joshdholtz — CHANGES_REQUESTED** [review]
2FA 😇

File: spaceship/lib/spaceship/client.rb
Comment (lacostej) [reply spaceship/lib/spaceship/client.rb]: If you get https://developer.apple.com/account/, you get a 302 to the login page you hardcoded.

Comment (youtux) [reply spaceship/lib/spaceship/client.rb]: Does this mean we can’t use passwords that contain `=` or `&`?

**Comment by taquitos** [comment]
I'm going to hold off on merging this until:

- We fix urlencoding of username/password
- 2FA is fixed

2FA is important because if we ship this right now, it will break everybody with it 😢 
It's a bit complicated though because we have to switch over from ITC 2FA to the Apple SSO 2FA which is significantly different. I'm about 50% of the way there, but it could take another day.

File: spaceship/lib/spaceship/two_step_client.rb
Comment (casz) [inline spaceship/lib/spaceship/two_step_client.rb]: why 🍎, would go away from a JSON post
kek https://github.com/amyworrall/QuickRadar/blob/8326a52ba49f47c1faad59117ead6935987f91e0/QuickRadar/QRRadarSubmissionService.m#L247
https://github.com/clayfreeman/IDMSWebAuth/blob/f5e3a6ed8be4bc76f53f7280209faaaa8d40823d/include/IDMSWebAuth.php#L147
seems like you should handle between 4 or 6 digit.

Could we do some kind of array and then just join on '&'

File: spaceship/lib/spaceship/client.rb
Comment (lacostej) [inline spaceship/lib/spaceship/client.rb]: @taquitos WRT to the hardcoded appID, could we add this as documentation?

# GET https://developer.apple.com/account/ returns a 302 redirect to idmsa.apple.com...

Comment (lacostej) [reply spaceship/lib/spaceship/client.rb]: Also there was a question from @youtux WRT encoding passwords.

https://github.com/fastlane/fastlane/pull/10826/files/8bf8bb3d3c4232356a035d4e8209756ec6a70c1f#r149506222

File: spaceship/lib/spaceship/client.rb
Comment (casz) [reply spaceship/lib/spaceship/client.rb]: I assume we have to url encode the password

```ruby
require 'uri'
URI.encode(password)
```

Comment (youtux) [reply spaceship/lib/spaceship/client.rb]: and the username as well

Comment (taquitos) [reply spaceship/lib/spaceship/client.rb]: Thank you, updated!

File: spaceship/lib/spaceship/two_step_client.rb
Comment (joshdholtz) [inline spaceship/lib/spaceship/two_step_client.rb]: @taquitos Should we uri encode `device_id` and `scnt` as well? I don't know _exactly_ what these values look like but better safe than sorry if we don't have full control over this input?

File: spaceship/lib/spaceship/two_step_client.rb
Comment (joshdholtz) [inline spaceship/lib/spaceship/two_step_client.rb]: Insert another dumb uri encode question here 😇

**Review by joshdholtz — COMMENTED** [review]
A few thoughts on this great work of yours 💪

**Comment by taquitos** [comment]
Posting for visibility, I updated the issue: https://github.com/fastlane/fastlane/issues/10793#issuecomment-342906293

File: spaceship/lib/spaceship/two_step_client.rb
Comment (casz) [inline spaceship/lib/spaceship/two_step_client.rb]: Should the 4 digit not be updated?

File: spaceship/lib/spaceship/client.rb
Comment (casz) [inline spaceship/lib/spaceship/client.rb]: for readability could we put the apiIdKey into a variable?

**Comment by taquitos** [comment]
While the PR fixes this for the simple login case, it only addresses SMS 2FA, and it employs web scraping, which is super flakey. 

Furthermore, the authentication logic we already have addresses so many corner cases, and is battle tested. Replacing that with this flakey update could cause a lot of pain for everybody.

Instead, we think the right thing moving forward is to assume Apple isn’t going to fix this obvious issue, and then focus our energy on improving our session caching, so that between various actions, we don’t login multiple times unless explicitly called for.

This will take some time to properly address, but we’ll have a plan soon and we’ll share that out as soon as we do 🚀. In the meantime, we’re going to keep this PR alive if it is helping to unblock people.

We have some workarounds whose success seem to vary:

- Login using a different account from same IP
- Change IP address (VPN/Proxy)
- Login from a different IP, it unblocks your original IP.
- Login in developer.apple.com and then navigate to itunesconnect.apple.com

File: spaceship/lib/spaceship/client.rb
Comment (bolom) [reply spaceship/lib/spaceship/client.rb]: it's possible to use
encoded_password = CGI.escape(password)
 encoded_apple_id = CGI.escape(user)

when you have "+" in user. it's not working

File: spaceship/lib/spaceship/client.rb
Comment (KrauseFx) [reply spaceship/lib/spaceship/client.rb]: I don't think we're actually allowed to have that API key as part of the fastlane code base

**Comment by KrauseFx** [comment]
Hey everyone, thanks again for looking into this issue. Sorry it took so long, I finally found some time to look further into this issue, and was able to reproduce the issue and prepare a fix. You can check out the PR over here https://github.com/fastlane/fastlane/pull/11108

I'd love to ship this as a new release tomorrow morning 👍

**Pull request closed without merging**
